README.md
# mimeconvertor README

Mime tools for Visual Studio Code. Inspired from Notepad++'s mime tools.

## Features

Convert text to MD5

Base 64 Encode Text

Base 64 Decode Text

URL Encode Text

URL Decode Text

Full URL Encode Text

Convert text to Lowercase

Convert text to Uppercase

Convert text to Capitalize

Convert Snake case to Camel case (e.g. this_is_snake_case --> thisIsCamelCase)

## Requirements

Not exactly. For md5 it works on native node crypto module

## Extension Settings

Not any available now

## Known Issues

Dont know any. If found raise an issue in github.

## Release Notes

### 1.0
Initial release

-----------------------------------------------------------------------------------------------------------

## LICENSE

MIT License, see License



CHANGELOG.md
# Change Log
0.1.0 - Initial Release

0.1.1 - Added git repo

src/main.ts
export var Mime = {

    _k: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",

    e64: function (_v: any) { // base 64 encode
        let _r: string = '';
        let _c1, _c2, _c3, _e1, _e2, _e3, _e4;
        let i = 0, _l = _v.length;

        while (i < _l) {

            _c1 = _v.charCodeAt(i++);
            _c2 = _v.charCodeAt(i++);
            _c3 = _v.charCodeAt(i++);

            _e1 = _c1 >> 2;
            _e2 = ((_c1 & 3) << 4) | (_c2 >> 4);
            _e3 = ((_c2 & 15) << 2) | (_c3 >> 6);
            _e4 = _c3 & 63;

            if (isNaN(_c2)) {
                _e3 = _e4 = 64;
            } else if (isNaN(_c3)) {
                _e4 = 64;
            }

            _r = _r +
                this._k.charAt(_e1) + this._k.charAt(_e2) +
                this._k.charAt(_e3) + this._k.charAt(_e4);
        }

        return _r;
    },

    d64: function (_v: any) { // base 64 decode
        let _r:string = "";
        let _c1, _c2, _c3, _e1, _e2, _e3, _e4;
        let i = 0, _l = _v.length;

        _v = _v.replace(/[^A-Za-z0-9\+\/\=]/g, "");

        while (i < _l) {

            _e1 = this._k.indexOf(_v.charAt(i++));
            _e2 = this._k.indexOf(_v.charAt(i++));
            _e3 = this._k.indexOf(_v.charAt(i++));
            _e4 = this._k.indexOf(_v.charAt(i++));

            _c1 = (_e1 << 2) | (_e2 >> 4);
            _c2 = ((_e2 & 15) << 4) | (_e3 >> 2);
            _c3 = ((_e3 & 3) << 6) | _e4;

            _r = _r + String.fromCharCode(_c1);

            if (_e3 !== 64) {
                _r = _r + String.fromCharCode(_c2);
            }
            if (_e4 !== 64) {
                _r = _r + String.fromCharCode(_c3);
            }

        }

        return _r;

    },

    utf8e: function (string: any) { // utf8 encode
        string = string.replace(/\r\n/g, "\n");
        let _r = "";

        for (let n = 0; n < string.length; n++) {

            let c = string.charCodeAt(n);

            if (c < 128) {
                _r += String.fromCharCode(c);
            }
            else if ((c > 127) && (c < 2048)) {
                _r += String.fromCharCode((c >> 6) | 192);
                _r += String.fromCharCode((c & 63) | 128);
            }
            else {
                _r += String.fromCharCode((c >> 12) | 224);
                _r += String.fromCharCode(((c >> 6) & 63) | 128);
                _r += String.fromCharCode((c & 63) | 128);
            }

        }

        return _r;
    },

    utf8d: function (utftext: any) { // utf8 decode
        let _r = "", i = 0, c = 0, c2 = 0, c3 = 0;

        while (i < utftext.length) {

            c = utftext.charCodeAt(i);

            if (c < 128) {
                _r += String.fromCharCode(c);
                i++;
            }
            else if ((c > 191) && (c < 224)) {
                c2 = utftext.charCodeAt(i + 1);
                _r += String.fromCharCode(((c & 31) << 6) | (c2 & 63));
                i += 2;
            }
            else {
                c2 = utftext.charCodeAt(i + 1);
                c3 = utftext.charCodeAt(i + 2);
                _r += String.fromCharCode(((c & 15) << 12) | ((c2 & 63) << 6) | (c3 & 63));
                i += 3;
            }

        }

        return _r;
    },

    eurl: function (_v: any) { // url encode
        return encodeURI(_v);
    },

    durl: function (_v: any) { // url decode
        return decodeURI(_v);
    },

    eurlf: function (_v: any) { // full url encode
        return encodeURIComponent(_v);
    },

    lcase: function (_v: any) { // lower case
        return _v.toLowerCase();
    },

    ucase: function (_v: any) { // uppercase
        return _v.toUpperCase();
    },

    ccase: function (_v: any) { // capitalize case
        return _v.split(' ').map( (w:string) => w.charAt(0).toUpperCase() + w.slice(1)).join(' ');
    },

    s2c: function (_v: any) {   // snake to camel case
        return _v.replace(/(\_\w)/g, function (m: any) { return m[1].toUpperCase(); });
    },

    random: function (_v: any) {
        // return window.crypto.getRandomValues(array);
        return _v;
    }

};


src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import { Mime } from './main';
import { createHash } from 'crypto';

// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {


    let encode64 = vscode.commands.registerCommand('extension.encode64', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.e64(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let decode64 = vscode.commands.registerCommand('extension.decode64', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.d64(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let encodemd5 = vscode.commands.registerCommand('extension.encodemd5', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = createHash('md5').update(text).digest('hex');

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let encodeutf8 = vscode.commands.registerCommand('extension.encodeutf8', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.utf8e(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let decodeutf8 = vscode.commands.registerCommand('extension.decodeutf8', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.utf8d(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let encodeurl = vscode.commands.registerCommand('extension.encodeurl', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.eurl(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let decodeurl = vscode.commands.registerCommand('extension.decodeurl', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.durl(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let fullencodeurl = vscode.commands.registerCommand('extension.fullencodeurl', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.eurlf(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let lcase = vscode.commands.registerCommand('extension.lcase', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.lcase(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let ucase = vscode.commands.registerCommand('extension.ucase', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.ucase(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let ccase = vscode.commands.registerCommand('extension.ccase', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.ccase(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });

    let s2c = vscode.commands.registerCommand('extension.s2c', () => {

        let editor = vscode.window.activeTextEditor;
        if (!editor) {
            return; // No open text editor
        }

        let selection = editor.selection;
        let text = editor.document.getText(selection);
        let response = Mime.s2c(text);

        editor.edit(function (editBuilder) {
            editBuilder.replace(selection, response);
        });

    });


    context.subscriptions.push(encodemd5);
    context.subscriptions.push(encode64);
    context.subscriptions.push(decode64);
    context.subscriptions.push(encodeutf8);
    context.subscriptions.push(decodeutf8);
    context.subscriptions.push(encodeurl);
    context.subscriptions.push(decodeurl);
    context.subscriptions.push(fullencodeurl);
    context.subscriptions.push(lcase);
    context.subscriptions.push(ucase);
    context.subscriptions.push(ccase);
    context.subscriptions.push(s2c);
}

// this method is called when your extension is deactivated
export function deactivate() {
}

src/md5.ts
/* from https://github.com/jbt/js-crypto */
/* 
export let md5 = function () {

    var k: any = [], i = 0;

    for (; i < 64;) {
        k[i] = 0 | (Math.abs(Math.sin(++i)) * 4294967296);
    }

    function calcMD5(str: any) {
        var b, c, d, j,
            x: any = [],
            str2 = unescape(encodeURI(str)),
            a = str2.length,
            h = [b = 1732584193, c = -271733879, ~b, ~c],
            i = 0;

        for (; i <= a;){
            x[i >> 2] |= (str2.charCodeAt(i) || 128) << 8 * (i++ % 4);
        }

        x[str = (a + 8 >> 6) * 16 + 14] = a * 8;
        i = 0;

        for (; i < str; i += 16) {
            let a:any = h; j = 0;
            for (; j < 64;) {
                a = [
                    d = a[3],
                    ((b = a[1] | 0) +
                        ((d = (
                            (a[0] +
                                [
                                    b & (c = a[2]) | ~b & d,
                                    d & b | ~d & c,
                                    b ^ c ^ d,
                                    c ^ (b | ~d)
                                ][a = j >> 4]
                            ) +
                            (k[j] +
                                (x[[
                                    j,
                                    5 * j + 1,
                                    3 * j + 5,
                                    7 * j
                                ][a] % 16 + i] | 0)
                            )
                        )) << (a = [
                            7, 12, 17, 22,
                            5, 9, 14, 20,
                            4, 11, 16, 23,
                            6, 10, 15, 21
                        ][4 * a + j++ % 4]) | d >>> 32 - a)
                    ),
                    b,
                    c
                ];
            }
            for (j = 4; j;){
                h[--j] = h[j] + a[j];
            }
        }

        str = '';
        for (; j < 32;){
            str += ((h[j >> 3] >> ((1 ^ j++ & 7) * 4)) & 15).toString(16);
        }
        return str;
    }
    return calcMD5;
}();
 */

/**
*
*  MD5 (Message-Digest Algorithm)
*  http://www.webtoolkit.info/
*
**/
export let md5 = function (string: string) {

    function RotateLeft(lValue: any, iShiftBits: any) {
        return (lValue << iShiftBits) | (lValue >>> (32 - iShiftBits));
    }

    function AddUnsigned(lX: any, lY: any) {
        var lX4, lY4, lX8, lY8, lResult;
        lX8 = (lX & 0x80000000);
        lY8 = (lY & 0x80000000);
        lX4 = (lX & 0x40000000);
        lY4 = (lY & 0x40000000);
        lResult = (lX & 0x3FFFFFFF) + (lY & 0x3FFFFFFF);
        if (lX4 & lY4) {
            return (lResult ^ 0x80000000 ^ lX8 ^ lY8);
        }
        if (lX4 | lY4) {
            if (lResult & 0x40000000) {
                return (lResult ^ 0xC0000000 ^ lX8 ^ lY8);
            } else {
                return (lResult ^ 0x40000000 ^ lX8 ^ lY8);
            }
        } else {
            return (lResult ^ lX8 ^ lY8);
        }
    }

    function F(x: any, y: any, z: any) { return (x & y) | ((~x) & z); }
    function G(x: any, y: any, z: any) { return (x & z) | (y & (~z)); }
    function H(x: any, y: any, z: any) { return (x ^ y ^ z); }
    function I(x: any, y: any, z: any) { return (y ^ (x | (~z))); }

    function FF(a: any, b: any, c: any, d: any, x: any, s: any, ac: any) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(F(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    }

    function GG(a: any, b: any, c: any, d: any, x: any, s: any, ac: any) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(G(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    }

    function HH(a: any, b: any, c: any, d: any, x: any, s: any, ac: any) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(H(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    }

    function II(a: any, b: any, c: any, d: any, x: any, s: any, ac: any) {
        a = AddUnsigned(a, AddUnsigned(AddUnsigned(I(b, c, d), x), ac));
        return AddUnsigned(RotateLeft(a, s), b);
    }

    function ConvertToWordArray(string: string) {
        var lWordCount;
        var lMessageLength = string.length;
        var lNumberOfWords_temp1 = lMessageLength + 8;
        var lNumberOfWords_temp2 = (lNumberOfWords_temp1 - (lNumberOfWords_temp1 % 64)) / 64;
        var lNumberOfWords = (lNumberOfWords_temp2 + 1) * 16;
        var lWordArray = Array(lNumberOfWords - 1);
        var lBytePosition = 0;
        var lByteCount = 0;
        while (lByteCount < lMessageLength) {
            lWordCount = (lByteCount - (lByteCount % 4)) / 4;
            lBytePosition = (lByteCount % 4) * 8;
            lWordArray[lWordCount] = (lWordArray[lWordCount] | (string.charCodeAt(lByteCount) << lBytePosition));
            lByteCount++;
        }
        lWordCount = (lByteCount - (lByteCount % 4)) / 4;
        lBytePosition = (lByteCount % 4) * 8;
        lWordArray[lWordCount] = lWordArray[lWordCount] | (0x80 << lBytePosition);
        lWordArray[lNumberOfWords - 2] = lMessageLength << 3;
        lWordArray[lNumberOfWords - 1] = lMessageLength >>> 29;
        return lWordArray;
    }

    function WordToHex(lValue: any) {
        var WordToHexValue = "", WordToHexValue_temp = "", lByte, lCount;
        for (lCount = 0; lCount <= 3; lCount++) {
            lByte = (lValue >>> (lCount * 8)) & 255;
            WordToHexValue_temp = "0" + lByte.toString(16);
            WordToHexValue = WordToHexValue + WordToHexValue_temp.substr(WordToHexValue_temp.length - 2, 2);
        }
        return WordToHexValue;
    }

    function Utf8Encode(string: string) {
        string = string.replace(/\r\n/g, "\n");
        var utftext = "";

        for (var n = 0; n < string.length; n++) {

            var c = string.charCodeAt(n);

            if (c < 128) {
                utftext += String.fromCharCode(c);
            }
            else if ((c > 127) && (c < 2048)) {
                utftext += String.fromCharCode((c >> 6) | 192);
                utftext += String.fromCharCode((c & 63) | 128);
            }
            else {
                utftext += String.fromCharCode((c >> 12) | 224);
                utftext += String.fromCharCode(((c >> 6) & 63) | 128);
                utftext += String.fromCharCode((c & 63) | 128);
            }

        }

        return utftext;
    }

    var x = Array();
    var k, AA, BB, CC, DD, a, b, c, d;
    var S11 = 7, S12 = 12, S13 = 17, S14 = 22;
    var S21 = 5, S22 = 9, S23 = 14, S24 = 20;
    var S31 = 4, S32 = 11, S33 = 16, S34 = 23;
    var S41 = 6, S42 = 10, S43 = 15, S44 = 21;

    string = Utf8Encode(string);

    x = ConvertToWordArray(string);

    a = 0x67452301; b = 0xEFCDAB89; c = 0x98BADCFE; d = 0x10325476;

    for (k = 0; k < x.length; k += 16) {
        AA = a; BB = b; CC = c; DD = d;
        a = FF(a, b, c, d, x[k + 0], S11, 0xD76AA478);
        d = FF(d, a, b, c, x[k + 1], S12, 0xE8C7B756);
        c = FF(c, d, a, b, x[k + 2], S13, 0x242070DB);
        b = FF(b, c, d, a, x[k + 3], S14, 0xC1BDCEEE);
        a = FF(a, b, c, d, x[k + 4], S11, 0xF57C0FAF);
        d = FF(d, a, b, c, x[k + 5], S12, 0x4787C62A);
        c = FF(c, d, a, b, x[k + 6], S13, 0xA8304613);
        b = FF(b, c, d, a, x[k + 7], S14, 0xFD469501);
        a = FF(a, b, c, d, x[k + 8], S11, 0x698098D8);
        d = FF(d, a, b, c, x[k + 9], S12, 0x8B44F7AF);
        c = FF(c, d, a, b, x[k + 10], S13, 0xFFFF5BB1);
        b = FF(b, c, d, a, x[k + 11], S14, 0x895CD7BE);
        a = FF(a, b, c, d, x[k + 12], S11, 0x6B901122);
        d = FF(d, a, b, c, x[k + 13], S12, 0xFD987193);
        c = FF(c, d, a, b, x[k + 14], S13, 0xA679438E);
        b = FF(b, c, d, a, x[k + 15], S14, 0x49B40821);
        a = GG(a, b, c, d, x[k + 1], S21, 0xF61E2562);
        d = GG(d, a, b, c, x[k + 6], S22, 0xC040B340);
        c = GG(c, d, a, b, x[k + 11], S23, 0x265E5A51);
        b = GG(b, c, d, a, x[k + 0], S24, 0xE9B6C7AA);
        a = GG(a, b, c, d, x[k + 5], S21, 0xD62F105D);
        d = GG(d, a, b, c, x[k + 10], S22, 0x2441453);
        c = GG(c, d, a, b, x[k + 15], S23, 0xD8A1E681);
        b = GG(b, c, d, a, x[k + 4], S24, 0xE7D3FBC8);
        a = GG(a, b, c, d, x[k + 9], S21, 0x21E1CDE6);
        d = GG(d, a, b, c, x[k + 14], S22, 0xC33707D6);
        c = GG(c, d, a, b, x[k + 3], S23, 0xF4D50D87);
        b = GG(b, c, d, a, x[k + 8], S24, 0x455A14ED);
        a = GG(a, b, c, d, x[k + 13], S21, 0xA9E3E905);
        d = GG(d, a, b, c, x[k + 2], S22, 0xFCEFA3F8);
        c = GG(c, d, a, b, x[k + 7], S23, 0x676F02D9);
        b = GG(b, c, d, a, x[k + 12], S24, 0x8D2A4C8A);
        a = HH(a, b, c, d, x[k + 5], S31, 0xFFFA3942);
        d = HH(d, a, b, c, x[k + 8], S32, 0x8771F681);
        c = HH(c, d, a, b, x[k + 11], S33, 0x6D9D6122);
        b = HH(b, c, d, a, x[k + 14], S34, 0xFDE5380C);
        a = HH(a, b, c, d, x[k + 1], S31, 0xA4BEEA44);
        d = HH(d, a, b, c, x[k + 4], S32, 0x4BDECFA9);
        c = HH(c, d, a, b, x[k + 7], S33, 0xF6BB4B60);
        b = HH(b, c, d, a, x[k + 10], S34, 0xBEBFBC70);
        a = HH(a, b, c, d, x[k + 13], S31, 0x289B7EC6);
        d = HH(d, a, b, c, x[k + 0], S32, 0xEAA127FA);
        c = HH(c, d, a, b, x[k + 3], S33, 0xD4EF3085);
        b = HH(b, c, d, a, x[k + 6], S34, 0x4881D05);
        a = HH(a, b, c, d, x[k + 9], S31, 0xD9D4D039);
        d = HH(d, a, b, c, x[k + 12], S32, 0xE6DB99E5);
        c = HH(c, d, a, b, x[k + 15], S33, 0x1FA27CF8);
        b = HH(b, c, d, a, x[k + 2], S34, 0xC4AC5665);
        a = II(a, b, c, d, x[k + 0], S41, 0xF4292244);
        d = II(d, a, b, c, x[k + 7], S42, 0x432AFF97);
        c = II(c, d, a, b, x[k + 14], S43, 0xAB9423A7);
        b = II(b, c, d, a, x[k + 5], S44, 0xFC93A039);
        a = II(a, b, c, d, x[k + 12], S41, 0x655B59C3);
        d = II(d, a, b, c, x[k + 3], S42, 0x8F0CCC92);
        c = II(c, d, a, b, x[k + 10], S43, 0xFFEFF47D);
        b = II(b, c, d, a, x[k + 1], S44, 0x85845DD1);
        a = II(a, b, c, d, x[k + 8], S41, 0x6FA87E4F);
        d = II(d, a, b, c, x[k + 15], S42, 0xFE2CE6E0);
        c = II(c, d, a, b, x[k + 6], S43, 0xA3014314);
        b = II(b, c, d, a, x[k + 13], S44, 0x4E0811A1);
        a = II(a, b, c, d, x[k + 4], S41, 0xF7537E82);
        d = II(d, a, b, c, x[k + 11], S42, 0xBD3AF235);
        c = II(c, d, a, b, x[k + 2], S43, 0x2AD7D2BB);
        b = II(b, c, d, a, x[k + 9], S44, 0xEB86D391);
        a = AddUnsigned(a, AA);
        b = AddUnsigned(b, BB);
        c = AddUnsigned(c, CC);
        d = AddUnsigned(d, DD);
    }

    var temp = WordToHex(a) + WordToHex(b) + WordToHex(c) + WordToHex(d);

    return temp.toLowerCase();
};

